tests/unit_tests/test_cylindrical_mesh.py

```diff
@@ -0,0 +1,105 @@
+from itertools import product, permutations
+
+import openmc
+import numpy as np
+
+import pytest
+
+geom_size = 5
+
+@pytest.fixture()
+def model():
+    openmc.reset_auto_ids()
+
+    water = openmc.Material(name='water')
+    water.add_element('H', 2.0)
+    water.add_element('O', 1.0)
+    water.set_density('g/cc', 1.0)
+
+    rpp = openmc.model.RectangularParallelepiped(*([-geom_size, geom_size] * 3),
+                                                 boundary_type='vacuum')
+
+    cell = openmc.Cell(region=-rpp, fill=water)
+
+    geom = openmc.Geometry([cell])
+
+    source = openmc.Source()
+    source.space = openmc.stats.Point()
+    source.energy = openmc.stats.Discrete([10000], [1.0])
+
+    settings = openmc.Settings()
+    settings.particles = 2000
+    settings.batches = 10
+    settings.run_mode = 'fixed source'
+
+    # build
+    mesh = openmc.CylindricalMesh()
+    mesh.phi_grid = np.linspace(0, 2*np.pi, 21)
+    mesh.z_grid = np.linspace(-geom_size, geom_size, 11)
+    mesh.r_grid = np.linspace(0, geom_size, geom_size)
+
+    tally = openmc.Tally()
+
+    mesh_filter = openmc.MeshFilter(mesh)
+    tally.filters.append(mesh_filter)
+
+    tally.scores.append("flux")
+
+    tallies = openmc.Tallies([tally])
+
+    return openmc.Model(geometry=geom, settings=settings, tallies=tallies)
+
+def test_origin_read_write_to_xml(run_in_tmpdir, model):
+    """Tests that the origin attribute can be written and read back to XML
+    """
+    mesh = model.tallies[0].filters[0].mesh
+    mesh.origin = [0.1, 0.2, 0.3]
+    model.tallies.export_to_xml()
+
+    # read back
+    new_tallies = openmc.Tallies.from_xml()
+    new_tally = new_tallies[0]
+    new_mesh = new_tally.filters[0].mesh
+    np.testing.assert_equal(new_mesh.origin, mesh.origin)
+
+estimators = ('tracklength', 'collision')
+origins = set(permutations((-geom_size, 0, 0)))
+origins |= set(permutations((geom_size, 0, 0)))
+
+test_cases = product(estimators, origins)
+
+def label(p):
+    if isinstance(p, tuple):
+        return f'origin:{p}'
+    if isinstance(p, str):
+        return f'estimator:{p}'
+
+@pytest.mark.parametrize('estimator,origin', test_cases, ids=label)
+def test_offset_mesh(model, estimator, origin):
+    """Tests that the mesh has been moved based on tally results
+    """
+    mesh = model.tallies[0].filters[0].mesh
+    model.tallies[0].estimator = estimator
+    # move the center of the cylinder mesh upwards
+    mesh.origin = origin
+
+    sp_filename = model.run()
+
+    with openmc.StatePoint(sp_filename) as sp:
+        tally = sp.tallies[1]
+
+        # we've translated half of the cylinder mesh above the model,
+        # so ensure that half of the bins are populated
+        assert np.count_nonzero(tally.mean) == tally.mean.size / 2
+
+        # check that the half of the mesh that is outside of the geometry
+        # contains the zero values
+        mean = tally.mean.reshape(mesh.dimension[::-1]).T
```

Comment:
perhaps use `get_reshaped_data` with `expand_dims` here to make it clearer what we are indexing at least in the test?

Reply:
Sorry, is `get_reshaped_data` a method of `mesh`?

Reply:
It's a method on the `openmc.Tally` class. @eepeterson just made some nice updates to it to expand the filter dimensions correctly for a `MeshFilter`.

https://docs.openmc.org/en/latest/pythonapi/generated/openmc.Tally.html#openmc.Tally.get_reshaped_data